libs/librepcb/common/algorithm/airwiresbuilder.cpp

```diff
@@ -0,0 +1,198 @@
+}
+
+AirWiresBuilder::AirWires AirWiresBuilder::buildAirWires() noexcept {
+  // remember how many edges are already known as connected
+  uint connectedEdges = mEdges.size();
+
+  // determine additional edges between found points (candidates for airwires)
+  if (mPoints.size() == 2) {
+    mEdges.emplace_back(mPoints[0], mPoints[1], -1);
+  } else if (mPoints.size() == 3) {
+    // manually triangulate since it is easy and more stable than the
+    // delaunay-triangulation library
+    mEdges.emplace_back(mPoints[0], mPoints[1], -1);
+    mEdges.emplace_back(mPoints[1], mPoints[2], -1);
+    mEdges.emplace_back(mPoints[2], mPoints[0], -1);
+  } else if (mPoints.size() >= 3) {
+    // since delaunay-triangulation sometimes doesn't work well, add fallback
+    // edges to make sure at least all points are connected somehow
+    for (std::size_t i = 1; i < mPoints.size(); ++i) {
+      mEdges.emplace_back(mPoints[i - 1], mPoints[i], -1);
+    }
+
+    // now run delaunay triangulation to add additional edges
+    delaunay::Delaunay<qreal> del;
+    del.triangulate(mPoints);
+    mEdges.insert(mEdges.end(), del.getEdges().begin(), del.getEdges().end());
+  }
+
+  // determine weights of these new edges
+  for (uint i = connectedEdges; i < mEdges.size(); ++i) {
+    mEdges[i].weight = mEdges[i].p1.dist2(mEdges[i].p2);
+  }
+
+  // find airwires in list of edges
+  return AirWiresBuilder::kruskalMst();
+}
+
+/*******************************************************************************
+ *  Private Methods
+ ******************************************************************************/
+
+// adapted from horizon/kicad
+AirWiresBuilder::AirWires AirWiresBuilder::kruskalMst() noexcept {
+  unsigned int nodeNumber      = mPoints.size();
+  unsigned int mstExpectedSize = nodeNumber - 1;
+  unsigned int mstSize         = 0;
+  bool         ratsnestLines   = false;
+
+  // printf("mst nodes : %d edges : %d\n", mPoints.size(), mEdges.size () );
+  // The output
+  AirWires mst;
+
+  // Set tags for marking cycles
+  std::unordered_map<int, int> tags;
+  unsigned int                 tag = 0;
+
+  for (auto& node : mPoints) {
+    node.tag      = tag;
+    tags[node.id] = tag++;
+  }
+
+  // Lists of nodes connected together (subtrees) to detect cycles in the
+  // graph
+  std::vector<std::list<int>> cycles(nodeNumber);
+
+  for (unsigned int i = 0; i < nodeNumber; ++i) cycles[i].push_back(i);
+
+  // Kruskal algorithm requires edges to be sorted by their weight
+  std::sort(mEdges.begin(), mEdges.end(),
+            [](const delaunay::Edge<qreal>& a, const delaunay::Edge<qreal>& b) {
+              return a.weight > b.weight;
+            });
+
+  while (mstSize < mstExpectedSize && !mEdges.empty()) {
+    auto& dt = mEdges.back();
+
+    int srcTag = tags[dt.p1.id];
+    int trgTag = tags[dt.p2.id];
+    // printf("mstSize %d %d %f %d<>%d\n", mstSize, mstExpectedSize,
+    // dt.weight, srcTag, trgTag);
```

Comment:
Driveby comment (I didn't really look at the logic): There's quite some commented out debug code in here (more below). Do you want to keep that?

Reply:
Yeah I don't like such comments too, but as stated above the function, this function is copied from Horizon (which copied it from KiCad) and I'd just like to keep it as-is to keep the diff to the upstream code as small as possible :wink:

Reply:
Ah, good point, I missed that! :+1: